tfhe/src/safe_serialization.rs

```diff
@@ -0,0 +1,205 @@
+        .map_err(|err| err.to_string())?;
+
+    if deserialized_type != T::NAME {
+        return Err(format!(
+            "On deserialization, expected type {}, got type {}",
+            T::NAME,
+            deserialized_type
+        ));
+    }
+
+    options
+        .with_limit(serialized_size_limit)
+        .deserialize_from(&mut reader)
+        .map_err(|err| err.to_string())
+}
+
+/// Deserializes an object with [safe_deserialize] and checks than it is conformant with the given
+/// parameter set
+pub fn safe_deserialize_conformant<T: DeserializeOwned + Named + ParameterSetConformant>(
+    reader: impl std::io::Read,
+    serialized_size_limit: u64,
+    parameter_set: &T::ParameterSet,
+) -> Result<T, String> {
+    let deser: T = safe_deserialize(reader, serialized_size_limit)?;
+
+    if !deser.is_conformant(parameter_set) {
+        return Err(format!(
+            "Deserialized object of type {} not conformant with given parameter set",
+            T::NAME
+        ));
+    }
+
+    Ok(deser)
+}
+
+#[cfg(all(test, feature = "shortint"))]
+mod test {
+    use crate::safe_serialization::{safe_deserialize_conformant, safe_serialize};
+    use crate::shortint::parameters::{
+        PARAM_MESSAGE_2_CARRY_2_KS_PBS, PARAM_MESSAGE_3_CARRY_3_KS_PBS,
+    };
+    use crate::shortint::{gen_keys, Ciphertext, PBSParameters};
+
+    #[test]
+    fn safe_ser_ct() {
+        let (ck, _sk) = gen_keys(PARAM_MESSAGE_2_CARRY_2_KS_PBS);
+
+        let msg = 2_u64;
+
+        let ct = ck.encrypt(msg);
+
+        let mut buffer = vec![];
+
+        safe_serialize(&ct, &mut buffer, 1 << 40).unwrap();
+
+        assert!(safe_deserialize_conformant::<Ciphertext>(
+            buffer.as_slice(),
+            1 << 40,
+            &PBSParameters::PBS(PARAM_MESSAGE_3_CARRY_3_KS_PBS).to_shortint_conformance_param(),
+        )
+        .is_err());
+
+        let ct2 = safe_deserialize_conformant(
+            buffer.as_slice(),
+            1 << 40,
+            &PBSParameters::PBS(PARAM_MESSAGE_2_CARRY_2_KS_PBS).to_shortint_conformance_param(),
+        )
+        .unwrap();
+
+        let dec = ck.decrypt(&ct2);
+        assert_eq!(msg, dec);
+    }
+}
+
+#[cfg(all(test, feature = "integer"))]
+mod test2 {
+    use crate::conformance::ParameterSetConformant;
+    use crate::high_level_api::{generate_keys, ConfigBuilder};
+
+    use crate::prelude::{FheDecrypt, FheTryEncrypt};
+    use crate::safe_serialization::{safe_deserialize_conformant, safe_serialize};
+    use crate::shortint::parameters::{
+        PARAM_MESSAGE_2_CARRY_2_KS_PBS, PARAM_MESSAGE_3_CARRY_3_KS_PBS,
+    };
+    use crate::{CompactFheUint8, CompactPublicKey};
+
+    #[test]
+    fn safe_ser_ct() {
+        let config = ConfigBuilder::all_disabled()
+            .enable_default_integers()
+            .build();
+
+        let (client_key, _server_key) = generate_keys(config);
+
+        let public_key = CompactPublicKey::new(&client_key);
+
+        let msg = 27u8;
```

Comment:
randomize ?

Reply:
I prefer to make tests reproductible
Here, the decrypted value is not really what the test is about